reco_utils/recommender/sar/sar_singlenode.py

```diff
@@ -281,6 +284,63 @@ def score(self, test, remove_seen=False):
 
         return test_scores
 
+    def _get_top_k_items(self, users, scores, top_k, sort_top_k=False):
+        """Extract top K items from score matrix, optionally sort results per user
+
+        Args:
+            users (pd.Series): test users
+            scores (np.array): score matrix
+            top_k (int): number of top items to recommend
+            sort_top_k (bool): flag to sort top k results
+
+        Returns:
+            pd.DataFrame: top k recommendation items for each user
+        """
+
+        if self.n_items < top_k:
+            logger.warning(
+                "Number of items is less than top_k, limiting top_k to number of items"
+            )
+        k = min(top_k, self.n_items)
+
+        test_user_idx = np.arange(scores.shape[0])[:, None]
+
+        # get top K items and scores
+        logger.info("Getting top K")
+        # this determines the un-ordered top-k item indices for each user
+        top_items = np.argpartition(scores, -k, axis=1)[:, -k:]
+        top_scores = scores[test_user_idx, top_items]
+
+        if sort_top_k:
+            sort_ind = np.argsort(-top_scores)
+            top_items = top_items[test_user_idx, sort_ind]
+            top_scores = top_scores[test_user_idx, sort_ind]
+
+        df = pd.DataFrame(
+            {
+                self.col_user: np.repeat(users.drop_duplicates().values, k),
+                self.col_item: [
+                    self.index2item[item] for item in np.array(top_items).flatten()
+                ],
+                self.col_prediction: np.array(top_scores).flatten(),
+            }
+        )
+
+        # drop invalid items
+        return df.replace(-np.inf, np.nan).dropna()
+
+    def recommend_popular_items(self, top_k=10, sort_top_k=False):
```

Comment:
I don't understand very well this function and the reason why we are making get_topk_items private. Why is it only for user `test_users = pd.Series([0])`? also, don't we want to support the option of getting the top k items given some users?

FYI @anargyri @yueguoguo

Reply:
Don't we use `recommend_k_items()` in the SAR notebooks?

Reply:
Also `remove_seen` is a flag we need.

Reply:
Since the PR is about new functionality, I don't see why the old functionality and signature should change at all.

Reply:
Maybe pass `users` in `_get_top_k_items()` as a list or numpy array would have been simpler.

Reply:
this is just refactoring to avoid repeating the same functionality in all 3 recommend methods. maybe the name isn't great. alternatively we could adjust the input / output further and replace the get_top_k_items method in python_evaluation?

Reply:
since _get_top_k_items needs to support cases where there are multiple users I need to create a series for the users (test_users), this is kind of hacky. I suppose a better way would be to remove users from get_top_k, and handle the generation of the dataframe with outside of that function.

Reply:
I am not sure what you mean by "multiple users"; isn't this the standard case that was supported before?
I agree that https://github.com/Microsoft/Recommenders/blob/9d76d1c81b3650326988b61094959cbd96efe709/reco_utils/evaluation/python_evaluation.py#L685
looks similar and could be refactored so that we have a single public function in the common utilities. Also @miguelgfierro has a point about making `_get_top_k_items()` public, because there is no fucntional dependence on SAR any more (only the column names).

Reply:
We have `recommend_popular_items(self, top_k=10, sort_top_k=False)` and `recommend_k_items(self, test, top_k=10, sort_top_k=False, remove_seen=False)`. I have two questions:
1) what is the difference between both?
2) Can't we have recommend_popular_items as a specific case of recommend_k_items?

Reply:
@miguelgfierro  ```recommend_popular_items``` just gets the most frequently occurring items across all users. we could use it as a fallback case if all the items in ```recommend_similar_items``` are cold, or the user is cold in recommend_k_items. Either way it seems useful to maintain as it's own method in case the user wants to access that information or if the method is shared across other functions. happy to change the names of this if you think something else would be clearer

Reply:
Oh I see, yeah I would suggest to rename it to something like what you said: `recommend_most_popular_items` and then add `gets the most frequently occurring items across all users.` in the docstrings

Reply:
I am wondering, since `recommend_most_popular_items` functionality is useful more generally with cold users, regardless of SAR; what would be a good way to make it available in other scenarios too e.g. when they train an ALS or SVD model for warm users but want a most popular model for cold ones.

Reply:
we could certainly offer a utility to determine that based on a dataframe of user/items, that might be more useful than this version since I'm only looking at counts not rating values, which could promote frequently occurring low-ranked items above moderately occurring high-ranked items